cpp/src/decisiontree/levelalgo/levelfunc_classifier.cuh

```diff
@@ -60,9 +58,9 @@ void grow_deep_tree_classification(
   }
   size_t total_nodes = pow(2, (maxdepth + 1)) - 1;
 
-  std::vector<std::vector<int>> sparse_histstate;
-  sparse_histstate.resize(total_nodes, std::vector<int>(n_unique_labels));
-  sparse_histstate[0] = histvec;
+  unsigned int* h_parent_hist = tempmem->h_parent_hist->data();
+  unsigned int* h_child_hist = tempmem->h_child_hist->data();
+  memcpy(h_parent_hist, histvec.data(), n_unique_labels * sizeof(int));
```

Comment:
Why use the bare pointers instead of `std::copy_n(histvec.begin(), n_unique_labels, h_parent_hist)`? Seems nice to stick with the containers.

Reply:
There are multiple issues to remove containers std:vector<std::vector<int>> and replace with array.

1. Its a std vector of std vector in classification; which is not uniform in memory.
2. I want to allocate all memories using cuml HostAllocator, so its pinned. Otherwise memcpy from/to GPU causes all stream to halt as they will be issued on default stream
3. Its hard to call cudaHostRegister on vector of vector. Might take a lot to time and also will block the streams.

also we had memcpy to host on pinned buffer and then assignment as well before. this was redundant.  

Does this make sense ?

Reply:
Would this be a case for unified memory? Seems like we have a lot of bookkeeping to sync data between gpu and host here that could be automated. I believe unified memory can be attached to a particular stream.

Reply:
Yes i think we could do that. Also unified memory will move data in chucks. This can be slower here, because we explicitly know size of the memory to be moved. 
But it can reduce the code complexity to a great extent,

This can also go away when we remove all CPU part in the future.

Reply:
Great, let's think about this for the upcoming cleanup stage, no need to solve it today. Would be worth investing in simplifying even if it's a slight perf hit.